File: src/main/java/pl/mjedynak/idea/plugins/builder/writer/BuilderContext.java
package pl.mjedynak.idea.plugins.builder.writer;

import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiDirectory;
import java.util.Arrays;
import java.util.Objects;
import pl.mjedynak.idea.plugins.builder.psi.model.PsiFieldsForBuilder;

public record BuilderContext(
        Project project,
        PsiFieldsForBuilder psiFieldsForBuilder,
        PsiDirectory targetDirectory,
        String className,
        PsiClass psiClassFromEditor,
        String methodPrefix,
        boolean isInner,
        boolean hasButMethod,
        boolean useSingleField,
        boolean hasAddCopyConstructor) {
    @Override
    public int hashCode() {
        Object[] objects = {project, psiFieldsForBuilder, targetDirectory, className, psiClassFromEditor, methodPrefix};
        return Arrays.hashCode(objects);
    }

    @SuppressWarnings("EqualsIncompatibleType")
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof BuilderContext other)) {
            return false;
        }
        return Objects.equals(this.project, other.project)
                && Objects.equals(this.psiFieldsForBuilder, other.psiFieldsForBuilder)
                && Objects.equals(this.targetDirectory, other.targetDirectory)
                && Objects.equals(this.className, other.className)
                && Objects.equals(this.psiClassFromEditor, other.psiClassFromEditor)
                && Objects.equals(this.methodPrefix, other.methodPrefix);
    }
}

File: src/main/kotlin/io/github/kingg22/kobra/builder/writer/BuilderContext.kt
package io.github.kingg22.kobra.builder.writer

import com.intellij.openapi.project.Project
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiDirectory
import io.github.kingg22.kobra.builder.psi.model.PsiFieldsForBuilder

// @JvmRecord
public data class BuilderContext(
    public val project: Project,
    public val psiFieldsForBuilder: PsiFieldsForBuilder,
    public val targetDirectory: PsiDirectory?,
    public val className: String,
    @JvmField public val psiClassFromEditor: PsiClass,
    @JvmField public val methodPrefix: String?,
    @JvmField public val isInnerBuilder: Boolean,
    public val hasButMethod: Boolean,
    public val useSingleField: Boolean,
    public val hasAddCopyConstructor: Boolean,
) {
    override fun hashCode(): Int = arrayOf(
        project,
        psiFieldsForBuilder,
        targetDirectory,
        className,
        psiClassFromEditor,
        methodPrefix,
    ).contentHashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is BuilderContext) {
            return false
        }
        return this.project == other.project &&
            this.psiFieldsForBuilder == other.psiFieldsForBuilder &&
            this.targetDirectory == other.targetDirectory &&
            this.className == other.className &&
            this.psiClassFromEditor == other.psiClassFromEditor &&
            this.methodPrefix == other.methodPrefix
    }
}
